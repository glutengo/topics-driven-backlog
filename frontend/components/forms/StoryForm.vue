<template>
    <BForm 
        @submit.prevent="submit" 
        class="story-form">

        <BRow>
            <BCol md="10">
                <BFormGroup 
                    label="Story *" 
                    label-for="story-title">

                    <BFormInput 
                        id="story-title" 
                        v-model="data.title" 
                        required/>
                </BFormGroup>
            </BCol>
            <BCol md="2">
                <BFormGroup 
                    label="Story points" 
                    label-for="story-points">

                    <BFormInput 
                        id="story-points" 
                        type="number" 
                        min="0" 
                        v-model="data.points"/>
                </BFormGroup>
            </BCol>
        </BRow>

        <BFormGroup
                label="Topic"
                label-for="story-topic">

            <BFormSelect
                    id="story-topic"
                    v-model="data.topic_id">

                <option :value="null" selected>No topic</option>

                <option
                    v-for="topic in topics"
                    :value="topic.id"
                    :key="topic.id">

                    {{ topic.title }}
                </option>
            </BFormSelect>
        </BFormGroup>

        <BFormGroup 
            label="Notes" 
            label-for="story-description">

            <BFormTextarea 
                id="story-description" 
                v-model="data.description"/>
        </BFormGroup>

        <BButton 
            type="submit" 
            variant="primary"
            :class="{ 'is-disabled': isInvalid}">Save</BButton>

        <BButton 
            type="button" 
            variant="grey" 
            @click="cancel">Cancel</BButton>
    </BForm>
</template>

<script>
import { required, numeric, minValue } from 'vuelidate/lib/validators';
import BForm from '@bootstrap/form/form';
import BFormGroup from '@bootstrap/form-group/form-group';
import BFormInput from '@bootstrap/form-input/form-input';
import BFormTextarea from '@bootstrap/form-textarea/form-textarea';
import BFormSelect from '@bootstrap/form-select/form-select';
import BRow from '@bootstrap/layout/row';
import BCol from '@bootstrap/layout/col';
import BButton from '@bootstrap/button/button';
import BaseForm from '@/components/forms/BaseForm';

export default {
    name: 'StoryForm',
    components: { BForm, BFormGroup, BFormInput, BFormTextarea, BFormSelect, BRow, BCol, BButton },
    extends: BaseForm,
    props: {
        project: {
            type: Object,
            default: null,
        },
    },
    computed: {
        topics() {
            return this.project
                ? this.$store.getters['topics/all'](this.project.course_id)
                : [];
        },
    },
    validations: {
        data: {
            title: { required },
            points: { numeric, minValue: minValue(0) },
        },
    },
};
</script>