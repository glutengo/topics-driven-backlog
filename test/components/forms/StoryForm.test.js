import { mount, createLocalVue } from 'vue-test-utils';
import Vuelidate from 'vuelidate';
import StoryForm from '@/components/forms/StoryForm';

const localVue = createLocalVue();
localVue.use(Vuelidate);

describe('StoryForm.test.js', () => {
    let cmp;

    beforeEach(() => {
        cmp = mount(StoryForm, {
            localVue,
            propsData: {
                data: {
                    title: 'Test Story',
                    points: 5,
                },
            },
        });
    });

    it('has the expected html structure', () => {
        expect(cmp.element).toMatchSnapshot();
    });
});