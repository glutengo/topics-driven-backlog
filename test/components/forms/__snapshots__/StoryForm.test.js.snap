// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryForm.test.js has the expected html structure 1`] = `
<form
  class="story-form"
>
  <div
    class="row"
  >
    <div
      class="col-md-10"
    >
      <div
        aria-labelledby="__BVID__9___BV_label_"
        class="b-form-group form-group"
        id="__BVID__9_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="story-title"
          id="__BVID__9___BV_label_"
        >
          Story *
        </label>
        <div
          class=""
        >
          <input
            aria-required="true"
            class="form-control"
            id="story-title"
            required="required"
            type="text"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
    </div>
     
    <div
      class="col-md-2"
    >
      <div
        aria-labelledby="__BVID__11___BV_label_"
        class="b-form-group form-group"
        id="__BVID__11_"
        role="group"
      >
        <label
          class="col-form-label pt-0"
          for="story-points"
          id="__BVID__11___BV_label_"
        >
          Story points
        </label>
        <div
          class=""
        >
          <input
            class="form-control"
            id="story-points"
            min="0"
            type="number"
          />
          <!---->
          <!---->
          <!---->
        </div>
      </div>
    </div>
  </div>
   
  <div
    aria-labelledby="__BVID__13___BV_label_"
    class="b-form-group form-group"
    id="__BVID__13_"
    role="group"
  >
    <label
      class="col-form-label pt-0"
      for="story-description"
      id="__BVID__13___BV_label_"
    >
      Notes
    </label>
    <div
      class=""
    >
      <textarea
        class="form-control"
        id="story-description"
        rows="1"
        wrap="soft"
      />
      <!---->
      <!---->
      <!---->
    </div>
  </div>
   
  <button
    class="btn btn-primary"
    type="submit"
  >
    Save
  </button>
   
  <button
    class="btn btn-grey"
    type="button"
  >
    Cancel
  </button>
</form>
`;
